json_socket.md

```diff
@@ -0,0 +1,159 @@
+
+Multiple commands can be send using the same TCP connection. Invalid commands -
+e.g. invalid JSON syntax - will result in error responses, but the connection
+will not be terminated.
+
+## send message
+
+Sending a message to one recipient or one group. Attachments can be attached as
+as base64 encoded string.
+`recipient` and `groupId` must not be provided at the same time.
+
+### Request `send_message`
+```JSON
+{
+    "command": "send_message",
+    "reqId": "[request ID (optional)]",
+
+    "recipient": "[phone number]",
+    "groupId": "[base64 group ID]",
+
+    "dataMessage": {
+        "message": "[text message (optional)]",
+        "attachments": [{
+            "base64Data": "[base64 encoded data]",
+            "filename": "[filename (optional)]"
+        }]
+    }
+}
+```
+
+### Response `send_message`
+```JSON
+{
+    "command": "send_message",
+    "reqId": "[referencing request ID]",
+
+    "statusCode": 0,
+    "timestamp": 1234567
+}
+```
+
+## send reaction
+
+Reacting to an existing message.
+`recipient` and `groupId` must not be provided at the same time.
+
+### Request `send_reaction`
+```JSON
+{
+    "command": "send_reaction",
+    "reqId": "[request ID (optional)]",
+
+    "recipient": "[phone number]",
+    "groupId": "[base64 group ID]",
+
+    "reaction": {
+        "emoji": "😀",
+        "author": "[phone number of original message]",
+        "remove": false,
+        "timestamp": 1234567
+    }
+}
+```
+
+### Response `send_reaction`
+```JSON
+{
+    "command": "send_reaction",
+    "reqId": "[referencing request ID]",
+
+    "statusCode": 0,
+    "timestamp": 1234567
+}
+```
+
+## receive messages
+
+Receiving incoming messages. Command waits for `timeout` milliseconds for new messages. Default `1000`.
+
+Attachments can be omitted in the command response with `ignoreAttachments`. Default `false`.
+
+### Request `receive_messages`
+```JSON
+{
+    "command": "receive_messages",
+    "reqId": "[request ID (optional)]",
+
+    "timeout": 1000,
+    "ignoreAttachments": false
+}
+```
+
+### Response `receive_messages`
+```JSON
+{
+    "command": "receive_messages",
+    "reqId": "[referencing request ID]",
+
+    "statusCode": 0,
+    "messages": [{
+        "timestamp": 1234567,
+        "sender": "[senders phone number]",
+        "body": "[text message]",
+        "attachments": [{
+            "base64Data": "[base64 encoded data]",
+            "filename": "[filename (optional)]"
+        }]
+    },{
+        "statusCode": -1,
+        "errorMessage": "[message parsing error]"
+    }]
+}
+```
+
+## Error response
+```JSON
+{
+    "command": "[requested command]",
+    "reqId": "[referencing request ID]",
+
+    "statusCode": -1,
+    "errorMessage": "[additional information]"
+}
+```
+
+# Example
+Running the client
+```
+signal-cli --username +436XXYYYZZZZ socket
+```
+
+Sending command with e.g. `socat`
+```
+echo "{command:'receive_messages'}" | socat -t 2 TCP:localhost:6789
```

Comment:
This doesn't seem to work:

`2021/04/03 16:21:50 socat[154] E exactly 2 addresses required (there are 1); use option "-h" for help`